treerec/tskit/tables.c

```diff
@@ -3419,57 +3412,57 @@ typedef struct {
     tsk_id_t *site_id_map;
 } table_sorter_t;
 
-static int
-cmp_site(const void *a, const void *b) {
-    const tsk_site_t *ia = (const tsk_site_t *) a;
-    const tsk_site_t *ib = (const tsk_site_t *) b;
-    /* Compare sites by position */
-    int ret = (ia->position > ib->position) - (ia->position < ib->position);
-    if (ret == 0) {
-        /* Within a particular position sort by ID.  This ensures that relative ordering
-         * of multiple sites at the same position is maintained; the redundant sites
-         * will get compacted down by clean_tables(), but in the meantime if the order
-         * of the redundant sites changes it will cause the sort order of mutations to
-         * be corrupted, as the mutations will follow their sites. */
-        ret = (ia->id > ib->id) - (ia->id < ib->id);
-    }
-    return ret;
-}
-
-static int
-cmp_mutation(const void *a, const void *b) {
-    const tsk_mutation_t *ia = (const tsk_mutation_t *) a;
-    const tsk_mutation_t *ib = (const tsk_mutation_t *) b;
-    /* Compare mutations by site */
-    int ret = (ia->site > ib->site) - (ia->site < ib->site);
-    if (ret == 0) {
-        /* Within a particular site sort by ID. This ensures that relative ordering
-         * within a site is maintained */
-        ret = (ia->id > ib->id) - (ia->id < ib->id);
-    }
-    return ret;
-}
-
-static int
-cmp_edge(const void *a, const void *b) {
-    const edge_sort_t *ca = (const edge_sort_t *) a;
-    const edge_sort_t *cb = (const edge_sort_t *) b;
-
-    int ret = (ca->time > cb->time) - (ca->time < cb->time);
-    /* If time values are equal, sort by the parent node */
-    if (ret == 0) {
-        ret = (ca->parent > cb->parent) - (ca->parent < cb->parent);
-        /* If the parent nodes are equal, sort by the child ID. */
-        if (ret == 0) {
-            ret = (ca->child > cb->child) - (ca->child < cb->child);
-            /* If the child nodes are equal, sort by the left coordinate. */
-            if (ret == 0) {
-                ret = (ca->left > cb->left) - (ca->left < cb->left);
-            }
-        }
-    }
-    return ret;
-}
+//The following 3 comparator functions are replaced by new bool based  ones in parallel_sort.cpp
```

Comment:
So, bool-based comparators are not actually quite the same, right?  The int comparators return -1, 0, or 1, where 0 means "compared equal"; the bool-based comparators don't have that option.  What are the implications?  Will the sort order be subtly different when using the new comparators?  That could introduce some very subtle bugs that would have us tearing our hair out.  Is this change absolutely necessary?  @petrelharp, thoughts?

Reply:
I guess since qsort() and std::sort() are not guaranteed to be stable, perhaps this is a non-issue; our replacement sorting does not have to guarantee stability either...?  I do note that since the single-threaded case will presumably continue to use the old sorting strategy, so we presumably don't want to comment out the int-based comparators, but instead just add new bool-based comparators as well...?

Reply:
Regarding the comparison operator, @bhaller, what's your opinion about the standard <=> C++ operator? Is it appropriate to consider here?

Reply:
> Regarding the comparison operator, @bhaller, what's your opinion about the standard <=> C++ operator? Is it appropriate to consider here?

I'm afraid you overestimate my C++ knowledge, @brnorris03; I didn't know any such thing as a <=> operator existed in C++ until I googled it just now.  :->  I can delve into that question if necessary, but at present I have no knowledge to contribute.